uaclient/daemon.py

```diff
@@ -0,0 +1,128 @@
+import logging
+import time
+
+from uaclient import actions, exceptions, lock, status, util
+from uaclient.clouds import AutoAttachCloudInstance
+from uaclient.clouds.gcp import UAAutoAttachGCPInstance
+from uaclient.clouds.identity import cloud_instance_factory
+from uaclient.config import UAConfig
+
+LOG = logging.getLogger("ua.daemon")
+
+
+def start():
+    try:
+        util.subp(["systemctl", "start", "ubuntu-advantage.service"])
+    except exceptions.ProcessExecutionError as e:
+        LOG.warning(e)
+
+
+def stop():
+    try:
+        util.subp(["systemctl", "stop", "ubuntu-advantage.service"])
+    except exceptions.ProcessExecutionError as e:
+        LOG.warning(e)
+
+
+def attempt_auto_attach(cfg: UAConfig, cloud: AutoAttachCloudInstance):
+    try:
+        with lock.SpinLock(
+            cfg=cfg, lock_holder="ua.daemon.attempt_auto_attach"
+        ):
+            actions.auto_attach(cfg, cloud)
+    except exceptions.LockHeldError as e:
+        LOG.error(e)
+        cfg.add_notice(
+            "",
+            status.NOTICE_DAEMON_AUTO_ATTACH_LOCK_HELD.format(
+                operation=e.lock_holder
+            ),
+        )
+        LOG.debug("Failed to auto attach")
+        return
+    except Exception as e:
+        LOG.exception(e)
+        cfg.add_notice("", status.NOTICE_DAEMON_AUTO_ATTACH_FAILED)
+        lock.clear_lock_file_if_present()
```

Comment:
Isn't the lock cleared by default after we exit the context manager ?

Reply:
It should be. the only case where it wouldn't is if `delete_cache_key` failed for some reason. Then this would try again. I was just following the pattern we have elsewhere where we call this function when we catch an exception. It is a `noop` if the lock was already cleared so it seemed harmless.

I can remove it if you think it is unnecessary though